hie-plugin-api/Haskell/Ide/Engine/Ghc.hs

```diff
@@ -130,131 +129,185 @@ srcErrToDiag df rfm se = do
   (diags, errs) <- processMsgs errMsgs
   return (Diagnostics diags, errs)
 
--- ---------------------------------------------------------------------
 
-myWrapper :: GM.IOish m
+-- | Run a Ghc action and capture any diagnostics and errors produced.
+captureDiagnostics :: (MonadIO m, GhcMonad m)
   => (FilePath -> FilePath)
-  -> GM.GmlT m ()
-  -> GM.GmlT m (Diagnostics, AdditionalErrs)
-myWrapper rfm action = do
+  -> m r
+  -> m (Diagnostics, AdditionalErrs, Maybe r)
+captureDiagnostics rfm action = do
   env <- getSession
-  diagRef <- liftIO $ newIORef mempty
+  diagRef <- liftIO $ newIORef $ Diagnostics mempty
   errRef <- liftIO $ newIORef []
   let setLogger df = df { log_action = logDiag rfm errRef diagRef }
-      setDeferTypedHoles = setGeneralFlag' Opt_DeferTypedHoles
+      unsetWErr df = unSetGeneralFlag' Opt_WarnIsError (df {fatalWarningFlags = ES.empty})
```

Comment:
Hmm.. not sure this should be set here. We should set it when we initialize the other flags?

Reply:
Need a comment anyway.